doc/rst/developer/chips/23.rst

```diff
@@ -375,243 +370,407 @@ How does this impact our current view of two-phase initialization?
-current type is fully initialized.  In this sense it plays
-just one of the two roles of super.init() in the current
-implementation.  The compiler will insert initializations for
-omitted fields if necessary.
+     proc init(r) {
+       super.init(r);
 
-The name for this method is unimportant for this proposal but, for
-concreteness, we will refer to this method as initDone().  Consider
-the previous hierarchy and the implementation of C.init() that
-includes a use of this.initDone().  Before this call the object has
-static/dynamic type B.  After this call the static/dynamic type is C.
-This makes it possible to invoke methods defined on C and to pass
-'this' to functions that accept an A, a B, or a C.  Dynamic dispatch
-will recognize that this object has type C.
+       foo();       // This is MyClass1.foo()
 
+       f = 5;
 
+       initDone();
 
+       foo();       // This is MyClass2.foo()
+     }
 
+     proc foo() {
+       a[2] = 20;
+     }
+   }
 
-Variation 2 :- postInit() with formals
-++++++++++++++++++++++++++++++++++++++
+   var c : MyClass1 = new MyClass2(1 .. 3);
 
-In the base proposal an override for postInit() does not accept any
-formals.  This means there is just one postInit() no matter how many
-overrides there are for init().  Assume this is C.postInit() for the
-previous hierarchy.
+   writeln(c);      // => {d = {1..3}, a = 0 20 30, f = 5}
 
-Any business logic that this method implements must be based solely on
-the fields defined by the static type C.  Method calls will
-dynamically dispatch based on the type D.
 
-We consider the option to provide multiple overloads for C.postInit().
-In this model a call to
+
+
+
+
+The postInit() method
++++++++++++++++++++++
+
+This proposal introduces a new method; postInit() e.g.
+
+.. code-block:: chapel
+
+   class MyClass1 {
+     var d : domain(1);
+     var a : [d] int;
+
+     proc init(r) {
+       d = r;
+     }
+
+     proc postInit() {
+       foo();
+     }
+
+     proc foo() {
+       a[3] = 30;
+     }
+   }
+
+   class MyClass2 : MyClass1 {
+     var f : int;
+
+     proc init(r) {
+       super.init(r);
+
+       f = 5;
+     }
+
+     proc foo() {
+       a[2] = 20;
+     }
+   }
+
+   var c : MyClass1 = new MyClass2(1 .. 3);
+
+   writeln(c);      // => {d = {1..3}, a = 0 20 0, f = 5}
+
+
+A use of the new expression for a class
 
 .. code-block:: chapel
 
-   var  x = new D(...args...)
+   var x = new MyClass1(...);
 
-is equivalent to
+acts as if it were equivalent to
 
 .. code-block:: chapel
 
-   var  myInstance = allocate(D);
+   var x = allocate(MyClass1);
+
+   x.init(...args...);
+   x.postInit();
+
+i.e. an instance is allocated and then the appropriate overload of the
+MyClass1.init() method is invoked.  When this method returns 'x' is a
+fully initialized instance of MyClass1.  Then the method postInit() is
+invoked.  If this call results in method calls with virtual overloads
+then the dispatch will be for objects of type MyClass1.
```

Comment:
Hi @noakesmichael, I'm curious about some of the details here.

Particularly, how ``postInit()``s work with a class hierarchy.  For example the ``A -> B -> C -> D``, as in the "Two-phase initialization" section above.

If only ``A`` defines a ``postInit()``, then ``var d = new D()`` will as above do ``d.init(); d.postInit()``, and ``d.postInit()`` will resolve to the ``A.postInit()``.

But if multiple levels define ``postInit()``s, how is control handed off between them?  Would ``B.postInit()`` have a call to ``super.postInit()``?  If ``B.postInit()`` was defined without such, would the compiler insert one, declare it an error, or would control just not reach ``A.postInit()``?

If ``A`` and ``C`` define ``postInit()``, would ``B`` be required to define one to pass control along?  Or would the compiler insert one?  Or would ``C.postInit()``'s call to ``super.postInit()`` just work out to calling ``A.postInit()``?

Or would no ``postInit()`` invoke super.postInit()?  I could image the compiler, upon seeing ``var d = new D()``, could, as it knows the static type D and its static type hierarchy, insert the appropriate independent ``postInit()`` calls in some defined order, instead of just one ``d.postInit()``?